lib/rules/no-shadow.js

```diff
@@ -54,9 +64,120 @@ module.exports = {
+         * @returns {boolean} `true` if the location is inside of the range of the node.
+         */
+        function isInRange(node, location) {
+            return node && node.range[0] <= location && location <= node.range[1];
+        }
+
+        /**
+         * Checks whether or not the nodes types include the node type.
+         * @param {ASTNode} node a node to get.
+         * @param {[string]} types an array of nodes types.
+         * @returns {boolean} True if nodes types include the node type.
+         */
+        function isNodeMatchesTypes(node, types) {
+            return types.includes(node.type);
+        }
+
+        /**
+         * Searches from the current node through its ancestry to find a matching node.
+         * @param {ASTNode} node a node to get.
+         * @param {MatchCallback} match a callback that checks whether or not the node verifies its condition or not.
+         * @param {StopSearchingCallback} stopSearching a callback that checks whether or not the searching should be stopped or not.
+         * @returns {ASTNode} the matching node.
+         */
+        function findSelfOrAncestor(node, match, stopSearching) {
+            let currentNode = node;
+
+            while (currentNode && !match(currentNode)) {
+                if (stopSearching && stopSearching(currentNode)) {
+                    return null;
+                }
+                currentNode = currentNode.parent;
+            }
+            return currentNode;
+        }
+
+        /**
+         * Finds function's outer scope.
+         * @param {Scope} scope Function's own scope.
+         * @returns {Scope} Function's outer scope.
+         */
+        function getOuterScope(scope) {
+            const upper = scope.upper;
+
+            if (upper.type === "function-expression-name") {
+                return upper.upper;
+            }
+            return upper;
+        }
+
+        /**
+         * Checks if a variable and a shadowedVariable have the same init pattern ancestor.
+         * @param {Object} variable a variable to check.
+         * @param {Object} shadowedVariable a shadowedVariable to check.
+         * @returns {boolean} Whether or not the variable and the shadowedVariable have the same init pattern ancestor.
+         */
+        function isInitPatternNode(variable, shadowedVariable) {
+            const outerDef = shadowedVariable.defs[0];
+            const { variableScope } = variable.scope;
+
+
+            if (!(FUN_NODES_TYPES.includes(variableScope.block.type) && getOuterScope(variableScope) === shadowedVariable.scope)) {
+                return false;
+            }
+
+            const fun = variableScope.block;
+            const { parent } = fun;
+
+            const callExpression = findSelfOrAncestor(
+                parent,
+                node => isNodeMatchesTypes(node, CALL_EXP_NODE_TYPE),
+                node => isNodeMatchesTypes(node, FUN_NODES_TYPES)
```

Comment:
We added this to treat the case of nested functions, this is an example:

`const x = foo(() => { const bar = () => { return x => {}; }; return bar; })`

I think we don’t need it anymore because now we check whether or not the outer scope of the shadowing variable equals the scope of the shadowed variable.